In [1]:
import pandas as pd 

In [2]:
import numpy as np

In [3]:
import warnings

In [4]:
warnings.filterwarnings('ignore')

In [5]:
road = pd.read_csv('uk_road_accident.csv')

In [6]:
road

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [7]:
road.describe().round()

Latitude  Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.0   660653.0              660679.0            660679.0
mean       53.0       -1.0                   1.0                 2.0
std         1.0        1.0                   1.0                 1.0
min        50.0       -8.0                   1.0                 1.0
25%        51.0       -2.0                   1.0                 1.0
50%        52.0       -1.0                   1.0                 2.0
75%        53.0       -0.0                   1.0                 2.0
max        61.0        2.0                  68.0                32.0

In [8]:
road.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

In [9]:
road['Accident_Severity']

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [10]:
road['Accident Date']

0           5/6/2019
1           2/7/2019
2         26-08-2019
3         16-08-2019
4           3/9/2019
             ...    
660674    18-02-2022
660675    21-02-2022
660676    23-02-2022
660677    23-02-2022
660678    28-02-2022
Name: Accident Date, Length: 660679, dtype: object

In [11]:
road['Latitude']

0         51.506187
1         51.495029
2         51.517715
3         51.495478
4         51.488576
            ...    
660674    57.374005
660675    57.232273
660676    57.585044
660677    57.214898
660678    57.575210
Name: Latitude, Length: 660679, dtype: float64

In [12]:
road['Light_Conditions']

0               Darkness - lights lit
1                            Daylight
2         Darkness - lighting unknown
3                            Daylight
4               Darkness - lights lit
                     ...             
660674                       Daylight
660675         Darkness - no lighting
660676                       Daylight
660677         Darkness - no lighting
660678                       Daylight
Name: Light_Conditions, Length: 660679, dtype: object

In [13]:
road['District Area']

0         Kensington and Chelsea
1         Kensington and Chelsea
2         Kensington and Chelsea
3         Kensington and Chelsea
4         Kensington and Chelsea
                   ...          
660674                  Highland
660675                  Highland
660676                  Highland
660677                  Highland
660678                  Highland
Name: District Area, Length: 660679, dtype: object

In [14]:
road['Longitude']

0        -0.209082
1        -0.173647
2        -0.210215
3        -0.202731
4        -0.192487
            ...   
660674   -3.467828
660675   -3.809281
660676   -3.862727
660677   -3.823997
660678   -3.895673
Name: Longitude, Length: 660679, dtype: float64

In [15]:
road['Number_of_Casualties']

0         1
1         1
2         1
3         1
4         1
         ..
660674    2
660675    1
660676    1
660677    1
660678    1
Name: Number_of_Casualties, Length: 660679, dtype: int64

In [16]:
road['Number_of_Vehicles']

0         2
1         2
2         3
3         4
4         2
         ..
660674    1
660675    1
660676    3
660677    2
660678    1
Name: Number_of_Vehicles, Length: 660679, dtype: int64

In [17]:
road['Road_Surface_Conditions']

0                  Dry
1          Wet or damp
2                  Dry
3                  Dry
4                  Dry
              ...     
660674             Dry
660675    Frost or ice
660676    Frost or ice
660677     Wet or damp
660678     Wet or damp
Name: Road_Surface_Conditions, Length: 660679, dtype: object

In [18]:
road['Road_Type']

0         Single carriageway
1         Single carriageway
2                        NaN
3         Single carriageway
4                        NaN
                 ...        
660674    Single carriageway
660675    Single carriageway
660676    Single carriageway
660677    Single carriageway
660678      Dual carriageway
Name: Road_Type, Length: 660679, dtype: object

In [19]:
road['Urban_or_Rural_Area']

0         Urban
1         Urban
2         Urban
3         Urban
4         Urban
          ...  
660674    Rural
660675    Rural
660676    Rural
660677    Rural
660678    Rural
Name: Urban_or_Rural_Area, Length: 660679, dtype: object

In [20]:
road['Weather_Conditions']

0            Fine no high winds
1         Raining no high winds
2                           NaN
3            Fine no high winds
4                           NaN
                  ...          
660674       Fine no high winds
660675       Fine no high winds
660676       Fine no high winds
660677       Fine no high winds
660678    Snowing no high winds
Name: Weather_Conditions, Length: 660679, dtype: object

In [21]:
road['Vehicle_Type']

0                                          Car
1                                          Car
2                        Taxi/Private hire car
3         Bus or coach (17 or more pass seats)
4                                Other vehicle
                          ...                 
660674                                     Car
660675                                     Car
660676                                     Car
660677                   Motorcycle over 500cc
660678                                     Car
Name: Vehicle_Type, Length: 660679, dtype: object

In [22]:
road.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [30]:
road['Latitude'] = road['Latitude'].fillna(road['Latitude'].mode()[0])

In [31]:
road['Longitude'] = road['Longitude'].fillna(road['Longitude'].mode()[0])

In [32]:
road['Road_Surface_Conditions'] = road['Road_Surface_Conditions'].fillna(road['Road_Surface_Conditions'].mode()[0])

In [33]:
road['Road_Type'] = road['Road_Type'].fillna(road['Road_Type'].mode()[0])

In [35]:
road['Urban_or_Rural_Area'] = road['Urban_or_Rural_Area'].fillna(road['Urban_or_Rural_Area'].mode()[0])

In [36]:
road['Weather_Conditions'] = road['Weather_Conditions'].fillna(road['Weather_Conditions'].mode()[0])

In [37]:
road.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [47]:
road['Accident_Severity'] = road['Accident_Severity'].astype('object')
road['Accident Date'] = road['Accident Date'].astype('object')
road['Light_Conditions'] = road['Light_Conditions'].astype('object')
road['District Area'] = road['District Area'].astype('object')
road['Road_Surface_Conditions'] = road['Road_Surface_Conditions'].astype('object')
road['Road_Type'] = road['Road_Type'].astype('object')
road['Urban_or_Rural_Area'] = road['Urban_or_Rural_Area'].astype('object')
road['Weather_Conditions'] = road['Weather_Conditions'].astype('object')
road['Vehicle_Type'] = road['Vehicle_Type'].astype('object')

In [48]:
road.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h2>20 Question with descriptive analysis </h2>

1. How many accident happen base on the data?

In [50]:
number_accident = len(road)
print(number_accident)

660679


Analysis : There are 660679 accident base on the data

2. What is the distribution of accident severities by vehicle type?

In [53]:
pd.crosstab(road['Vehicle_Type'], road['Accident_Severity'])

Accident_Severity                      Fatal  Serious  Slight
Vehicle_Type                                                 
Agricultural vehicle                      21      282    1644
Bus or coach (17 or more pass seats)     325     3373   22180
Car                                     6577    66461  424954
Data missing or out of range               0        0       6
Goods 7.5 tonnes mgw and over            216     2321   14770
Goods over 3.5t. and under 7.5t           67      857    5172
Minibus (8 - 16 passenger seats)          29      276    1671
Motorcycle 125cc and under               189     2031   13049
Motorcycle 50cc and under                 95     1014    6494
Motorcycle over 125cc and up to 500cc    105     1014    6537
Motorcycle over 500cc                    339     3457   21861
Other vehicle                             70      767    4800
Pedal cycle                                6       39     152
Ridden horse                               0        0       4
Taxi/Private hire car                    155     1771   11368
Van / Goods 3.5 tonnes mgw or under      467     4554   29139

Analysis:
 This shows how different types of vehicles are associated with different severity levels.

3.What is the correlation matrix for numerical accident features?

In [54]:
road[['Number_of_Casualties','Number_of_Vehicles']].corr()

Number_of_Casualties  Number_of_Vehicles
Number_of_Casualties              1.000000            0.228889
Number_of_Vehicles                0.228889            1.000000

Analysis: This shows correlation between casualties and vehicles in accidents.

4. How many accidents involved more than 2 vehicles?

In [55]:
len(road[road['Number_of_Vehicles'] > 2])

67897

Analysis:
This gives the count of multi-vehicle accidents.

5.How does accident severity vary by road surface condition?

In [56]:
pd.crosstab(road['Road_Surface_Conditions'], road['Accident_Severity'])

Accident_Severity        Fatal  Serious  Slight
Road_Surface_Conditions                        
Dry                       5790    61708  381049
Flood over 3cm. deep        23      152     842
Frost or ice               193     2007   16317
Snow                        35      565    5290
Wet or damp               2620    23785  160303

Analysis: This shows how road surface conditions impact accident severity.

6. What is the distribution of accident severities?

In [57]:
road['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

Analysis: This provides the distribution of accident severity levels (Slight, Serious, Fatal).

7. What is the average number of casualties per accident?

In [59]:
road['Number_of_Casualties'].mean().round()

np.float64(1.0)

Analysis: This shows the average casualties per accident across the dataset.

8. What is the maximum number of casualties in a single accident?

In [61]:
road['Number_of_Casualties'].max()

np.int64(68)

Analysis: This indicates the worst-case accident in terms of casualties.

9. What is the average number of vehicles per accident?

In [63]:
road['Number_of_Vehicles'].mean().round()

np.float64(2.0)

Analysis: This gives an idea of how many vehicles are typically involved in an accident.

10. What is the most common road surface condition during accidents?

In [64]:
road['Road_Surface_Conditions'].value_counts().idxmax()

'Dry'

Analysis: This shows which road surface condition occurs most frequently during accidents.

11. What is the distribution of accidents by road type?

In [65]:
road['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

Analysis: This provides insight into which road types experience more accidents.

12. How many accidents occurred in urban vs rural areas?

In [66]:
road['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

Analysis: This helps compare accident counts in urban vs rural areas.

13.What is the most common weather condition during accidents?

In [70]:
road['Weather_Conditions'].value_counts().idxmax()

'Fine no high winds'

Analysis: This identifies the most frequent weather condition during accidents.

14. What is the distribution of accidents by light conditions?

In [71]:
road['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

Analysis: This shows how light conditions (daylight, darkness) affect accident frequency.

15. What is the most common vehicle type involved in accidents?

In [72]:
road['Vehicle_Type'].value_counts().idxmax()

'Car'

Analysis: This shows which type of vehicle is most commonly involved in accidents.

16. How many accidents occurred each year?

In [73]:
pd.to_datetime(road['Accident Date'], errors='coerce').dt.year.value_counts().sort_index()

Accident Date
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

Analysis: This shows accident trends over different years.

17. What is the correlation between number of vehicles and casualties?

In [75]:
road[['Number_of_Vehicles','Number_of_Casualties']].corr()

Number_of_Vehicles  Number_of_Casualties
Number_of_Vehicles              1.000000              0.228889
Number_of_Casualties            0.228889              1.000000

Analysis: This checks if more vehicles involved lead to higher casualties.

18. Which district area has the highest number of accidents?

In [76]:
road['District Area'].value_counts().idxmax()

'Birmingham'

Analysis: This identifies the district with the most accidents.

19. Which district has the highest average casualties per accident?

In [77]:
road.groupby('District Area')['Number_of_Casualties'].mean().idxmax()

'Blaeu Gwent'

Analysis: This identifies the district with the most severe accidents on average.

20.  What is the proportion of accident severities?

In [78]:
road['Accident_Severity'].value_counts(normalize=True)

Accident_Severity
Slight     0.853366
Serious    0.133525
Fatal      0.013109
Name: proportion, dtype: float64

Analysis: This shows the proportion of different accident severities.